torch/testing/_internal/common_methods_invocations.py

```diff
@@ -4,14 +4,133 @@
+                 dtypesIfROCM=None,  # dtypes this function is expected to work with on ROCM
+                 has_out_kwarg=None,  # whether this function has an out kwarg
+                 decorators=None,  # decorators to apply to generated tests
+                 tests_to_skip=None):  # regular expressions describing tests to skip
+        self.name = name
+
+        self.dtypes = dtypes
+        self.dtypesIfCPU = dtypesIfCPU
+        self.dtypesIfCUDA = dtypesIfCUDA
+        self.dtypesIfROCM = dtypesIfROCM
+
+        self.method_variant = getattr(torch.Tensor, name) if hasattr(torch.Tensor, name) else None
+        inplace_name = name + "_"
+        self.inplace_variant = getattr(torch.Tensor, inplace_name) if hasattr(torch.Tensor, name) else None
+        self.has_out_kwarg = has_out_kwarg
+
+        self.decorators = decorators
+        self.tests_to_skip = tests_to_skip
+
+    # Returns the op (currently only supports torch.<op_name>)
+    def getOp(self):
+        return getattr(torch, self.name)
+
+    # Returns the method variant of the op (torch.Tensor.<op_name>)
+    # or None if the op has no method variant
+    def getMethod(self):
+        return self.method_variant
+
+    # Returns the inplace variant of the op (torch.Tensor.<op_name>_)
+    # or None if the op has no inplace variant
+    def getInplace(self):
+        return self.inplace_variant
+
+    # Returns True if the test should be skipped and False otherwise
+    def shouldSkip(self, test_name, *, device, dtype):
+        if self.tests_to_skip is None:
+            return False
+
+        if device is None and dtype is None:
+            full_name = test_name
+        elif dtype is None:
+            assert device is not None
+            full_name = test_name + "_" + str(torch.device(device).type)
+        else:
+            assert device is not None
+            assert dtype is not None
+            full_name = test_name + "_" + str(torch.device(device).type) + "_" + str(dtype).split('.')[1]
+
+        for regex in self.tests_to_skip:
+            if re.match(regex, full_name):
+                return True
+
+        return False
+
+
+# Metadata class for unary "universal functions (ufuncs)" that accept a single
+# tensor and have common properties like:
+#   - they are elementwise functions
+#   - the input shape is the output shape
+#   - they typically have method and inplace variants
+#   - they typically support the out kwarg
+#   - they typically have NumPy or SciPy references
+# See NumPy's universal function documentation
+# (https://numpy.org/doc/1.18/reference/ufuncs.html) for more details
+# about the concept of ufuncs.
+class UnaryUfuncMeta(OpMeta):
+    def __init__(self,
+                 name,  # the string name of the function
+                 *,
+                 dtypes=get_common_float_types(),
+                 dtypesIfCPU=get_common_float_and_complex_types_plus_bfloat16(),
```

Comment:
I imagine these defaults will apply to quite a few more ufuncs than just unary

Reply:
I hope so! Otherwise we'll want to create even more dtype getters.

Reply:
The implication is, if I have to continually copy paste lines 101-104, that seems like a bad outcome (even though some logic is factored into functions, which helper functions are called for CPU/CUDA/ROCm is still information that I'd prefer not to have duplicated everywhere.)

Reply:
I imagine this will become clearer when you add a second class of functions.

Reply:
One solution that I think @VitalyFedyunin and @gchanan have proposed is to create tools that probe functions' properties (like which dtypes they support) and create a data file (e.g. a .yaml file) that the database reads. This is an interesting idea for the future but will require additional thought and architecture. Unfortunately our functions are pretty heterogenous in their dispatch.